File: app/src/main/java/com/xnote/app/domain/text/NotePlainText.kt
package com.xnote.app.domain.text

import com.xnote.app.domain.document.DrawingBlock
import com.xnote.app.domain.document.ImageBlock
import com.xnote.app.domain.document.NoteDocument
import com.xnote.app.domain.document.StickerBlock
import com.xnote.app.domain.document.TableBlock
import com.xnote.app.domain.document.TextBlock
import com.xnote.app.domain.document.plainText
import com.xnote.app.domain.model.Note
import com.xnote.app.domain.model.NoteKind

// -- Type Definitions

data class VisibleTextStats(
    val characterCount: Int,
    val latinWordCount: Int,
)

// -- Constants

private val LatinWordRegex = Regex("[A-Za-z]+(?:['’][A-Za-z]+)*")

const val DefaultSummaryLength = 80

// -- Functions

fun extractPlainText(document: NoteDocument): String {
    val parts = ArrayList<String>()
    for (block in document.blocks) {
        when (block) {
            is TextBlock -> {
                val text = block.inlines.plainText()
                if (text.isNotEmpty()) parts += text
            }
            is TableBlock -> {
                for (row in block.rows) {
                    for (cell in row.cells) {
                        val text = cell.inlines.plainText()
                        if (text.isNotEmpty()) parts += text
                    }
                }
            }
            is ImageBlock, is StickerBlock, is DrawingBlock -> Unit
        }
    }
    return parts.joinToString("\n")
}

fun extractPlainText(note: Note): String = when (note.kind) {
    NoteKind.Rich -> note.document?.let(::extractPlainText).orEmpty()
    NoteKind.Markdown -> MarkdownVisibleText.extract(note.markdownText.orEmpty())
}

fun summarizePlainText(plainText: String, maxLength: Int = DefaultSummaryLength): String {
    val collapsed = plainText.replace(Regex("\\s+"), " ").trim()
    if (collapsed.length <= maxLength) return collapsed
    return collapsed.take(maxLength).trimEnd()
}

fun visibleTextStats(plainText: String): VisibleTextStats = VisibleTextStats(
    characterCount = countVisibleCharacters(plainText),
    latinWordCount = LatinWordRegex.findAll(plainText).count(),
)

fun visibleTextStats(note: Note): VisibleTextStats = visibleTextStats(extractPlainText(note))

fun countVisibleCharacters(text: String): Int {
    var count = 0
    var index = 0
    while (index < text.length) {
        val codePoint = text.codePointAt(index)
        if (!Character.isWhitespace(codePoint)) {
            count += 1
        }
        index += Character.charCount(codePoint)
    }
    return count
}

File: flutter/lib/domain/text/note_plain_text.dart
import '../document/note_block.dart';
import '../document/note_document.dart';
import '../markdown/markdown_visible_text.dart';
import '../model/note.dart';

// -- Type Definitions

final class VisibleTextStats {
  const VisibleTextStats({
    required this.characterCount,
    required this.latinWordCount,
  });

  final int characterCount;
  final int latinWordCount;

  // -- Functions

  @override
  bool operator ==(Object other) {
    return other is VisibleTextStats &&
        other.characterCount == characterCount &&
        other.latinWordCount == latinWordCount;
  }

  @override
  int get hashCode => Object.hash(characterCount, latinWordCount);
}

// -- Constants

const defaultSummaryLength = 80;
final _latinWordPattern = RegExp(r"[A-Za-z]+(?:['’][A-Za-z]+)*");

// -- Functions

String extractDocumentPlainText(NoteDocument document) {
  final parts = <String>[];
  for (final block in document.blocks) {
    switch (block) {
      case TextBlock():
        final text = plainText(block.inlines);
        if (text.isNotEmpty) {
          parts.add(text);
        }
      case TableBlock():
        for (final row in block.rows) {
          for (final cell in row.cells) {
            final text = plainText(cell.inlines);
            if (text.isNotEmpty) {
              parts.add(text);
            }
          }
        }
      case ImageBlock() || StickerBlock() || DrawingBlock():
        break;
    }
  }
  return parts.join('\n');
}

String extractNotePlainText(Note note) => switch (note.kind) {
      NoteKind.rich => extractDocumentPlainText(note.document!),
      NoteKind.markdown => extractMarkdownVisibleText(note.markdownText!),
    };

String summarizePlainText(
  String plainText, {
  int maximumLength = defaultSummaryLength,
}) {
  if (maximumLength < 0) {
    throw ArgumentError.value(maximumLength, 'maximumLength');
  }
  final collapsed = plainText.replaceAll(RegExp(r'\s+'), ' ').trim();
  if (collapsed.length <= maximumLength) {
    return collapsed;
  }
  return collapsed.substring(0, maximumLength).trimRight();
}

VisibleTextStats visibleTextStats(String plainText) => VisibleTextStats(
      characterCount: countVisibleCharacters(plainText),
      latinWordCount: _latinWordPattern.allMatches(plainText).length,
    );

VisibleTextStats noteVisibleTextStats(Note note) =>
    visibleTextStats(extractNotePlainText(note));

int countVisibleCharacters(String text) {
  var count = 0;
  for (final codePoint in text.runes) {
    final character = String.fromCharCode(codePoint);
    if (character.trim().isNotEmpty) {
      count += 1;
    }
  }
  return count;
}

Note withDerivedText(Note note) {
  final plainText = extractNotePlainText(note);
  final stats = visibleTextStats(plainText);
  return note.copyWith(
    visibleCharacterCount: stats.characterCount,
    latinWordCount: stats.latinWordCount,
    summary: summarizePlainText(plainText),
  );
}
